docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,223 @@
+
+Database migrations are complex and introduce downtime concerns. Here are [some examples](https://blog.rainforestqa.com/2014-06-27-zero-downtime-database-migrations) of zero-downtime Rails migration workflows.
+
+Deployments including database migrations must make special considerations:
+
+1.  Deployed code must be consistent with the currently deployed schema at any point in time.
+2.  Database or table locking must be minimized or eliminated to prevent service outages.
+
+The workflows which are effective at ensuring zero downtime migrations are typically multi-phased. For a user orchestrating a zero downtime migration deployment, it's likely the user needs to verify each deployment step discretely, with the option to abort and rollback after each phase.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks introduce problems with Rails migrations:
+
+1. The migration hook will execute once per pod in the deployment. This is a race and can result in concurrent migration attempts against a shared database. This burdens the hook author with coordinating the hooks or otherwise safeguarding the migration code to ensure it is safe for concurrent execution.
+  1. The uncoordinated aspect of these hooks is at also at odds with the general requirement to provide discrete step verification for the deployment.
+2. The migration hook will execute in the same pod as application containers, consuming resources allocated for the application.
+  1. This can cause instability, as it's unlikely the application pod resource allocation takes into account the temporarily increased requirements of a transient deployment step.
+
+###### Deployment hooks
+
+Deployment hooks satisfy this use case by providing a means to execute the hook only once per logical deployment. The hook is expressed as a run-once pod which provides the migration with its own resource allocation decoupled from the application.
+
+
+#### Use-case: Call a webhook
+
+Consider a Java application whose deployment should result in a webhook being called to notify some external system of the new deployment's availability.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks are inappropriate for this use case. The webhook should be called only once. Without a hook which coordinates with other replicated pod hook executions, the hook would:
+
+1. Be fired once per pod in the deployment.
+2. Be fired during the scale-up of the deployment rather than following the logical deployment as a whole.
+
+###### Deployment hooks
+
+A deployment hook configured to run following the scale-up of the deployment would satisfy the use case.
+
+
+## Proposed design
+
+Deployment hooks are implemented as run-once pods which can be executed at one or both of the following points during the deployment lifecycle:
```

Comment:
----- Original Message -----
> > +
> > +###### Container lifecycle hooks
> > +
> > +Container lifecycle hooks are inappropriate for this use case. The webhook
> > should be called only once. Without a hook which coordinates with other
> > replicated pod hook executions, the hook would:
> > +
> > +1. Be fired once per pod in the deployment.
> > +2. Be fired during the scale-up of the deployment rather than following
> > the logical deployment as a whole.
> > +
> > +###### Deployment hooks
> > +
> > +A deployment hook configured to run following the scale-up of the
> > deployment would satisfy the use case.
> > +
> > +
> > +## Proposed design
> > +
> > +Deployment hooks are implemented as run-once pods which can be executed at
> > one or both of the following points during the deployment lifecycle:
> 
> So, after talking with @danmcp and @abhgupta, we seemed to agree the
> following inputs might be satisfactory for most needs:
> 
> 1. A command to run (as a string array)
> 2. A named reference to a container within the deployment pod template

Defaults to the first container.

> 
> For the hook execution, a pod would be created using the command (1) and
> whatever image is used by (2). The assumption (limitation) being that your
> hook code is available in your deployed images.
> 
> If we had the ability to mount a script as a volume in a pod, we could
> support running the provided script as the entrypoint against the specified
> container image.

We keep talking about "mounting" but I don't think that's useful.  Snippet should be provided direct with a DC as part of the process definition, or you make it part of a custom deployment.

> 
> Three more scenarios came up:
> 
> 1. Running a command not provided by the deployment image
>   1. Would require a third input: image reference (to use instead of a
>   container name reference)

Use a custom deployment process or put the command in your image.

> 2. Running more than one hook per phase
>   1. Would require handling an array of hooks, which introduces more
>   complexity (especially around failure modes: how to handle partial
>   failures, etc.)

You can make the hook call other hooks.

> 3. A combination of 1 and 2: running multiple hooks per phase, where the
> command for some is provided by the deployment images, and the command for
> others is provided by some other image(s).

Yuck

> 
> We also discussed this:
> 
> >Provide snippet of bash to run inside deployment process pod
> 
> This scenario assumes we have the mounting capability (or some other way to
> get the script into the container), and also has some problems:
> 
> 1. The context for your script is going to be dependent on the strategy in
> use (e.g. our deployer image), which seems unlikely to have the necessary
> code/userspace for the hook

The necessary code / userspace is "osc", curl.

> 2. The hooks would be constrained to the resource allocations for the
> deployer pod, which are likely to be extremely minimal

Again, "osc", curl, a few others.

> 
> Thoughts?